### PR TITLE
Feature: Complete note service

### LINKED ISSUE #542: Services should not throw `QueryFailedError` if constraints are violated
Currently ,`notesService.createNote` throws a `QueryFailedError: SQLITE_CONSTRAINT: UNIQUE constraint failed: Notes.alias` if a note with the same alias already exists. We should either check the uniqueness ourselves or wrap the DB-specific error with something generic.

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
Note service

### Description
This PR completes the note service and adds an unit test.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
fixes #542

### CODE REVIEW

File: src/notes/notes.service.ts [hunk 1]
```diff
@@ -123,65 +177,163 @@ export class NotesService {
...
+  /**
+   * @async
+   * Update a notes permissions.
+   * @param {Note} note - the note
+   * @param {NotePermissionsUpdateDto} newPermissions - the permissions the not should be set to
+   * @return {Note} the note with the new permissions
+   * @throws {NotInDBError} there is no note with this id or alias
+   * @throws {PermissionsUpdateInconsistentError} the new permissions specify a user or group twice.
+   */
+  async updateNotePermissions(
```
Comment (InnayTool) [inline src/notes/notes.service.ts:216]: Should it only be possible to delete permissions at once?

**Review by InnayTool — CHANGES_REQUESTED** [review]

File: src/errors/errors.ts [hunk 2]
```diff
@@ -23,3 +27,7 @@ export class TokenNotValidError extends Error {
 export class TooManyTokensError extends Error {
   name = 'TooManyTokensError';
 }
+
+export class PermissionsUpdateInconsistentError extends Error {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/errors/errors.ts:31]: I'd leave it as is.

File: src/notes/notes.service.ts (hunk 1 above)
Comment (DerMolly) [reply src/notes/notes.service.ts:216]: this should be fixed now

File: src/notes/notes.service.spec.ts [hunk 3]
```diff
@@ -67,9 +76,800 @@ describe('NotesService', () => {
...
     expect(service).toBeDefined();
   });
+
+  describe('getUserNotes', () => {
+    describe('works', () => {
+      const user = User.create('hardcoded', 'Testy') as User;
+      const alias = 'alias';
+      const note = Note.create(user, alias);
+
+      it('with one note', async () => {
```
Comment (davidmehren) [inline src/notes/notes.service.spec.ts:100]: Title does not match content (no note is used)

File: src/notes/notes.service.spec.ts [hunk 4]
```diff
@@ -67,9 +76,800 @@ describe('NotesService', () => {
...
+      it('with multiple note', async () => {
+        jest.spyOn(noteRepo, 'find').mockResolvedValueOnce([note, note]);
+        const notes = await service.getUserNotes(user);
+        expect(notes).toEqual([note, note]);
+      });
+    });
+  });
+
+  describe('createNote', () => {
+    describe('works', () => {
```
Comment (davidmehren) [inline src/notes/notes.service.spec.ts:121]: Maybe move the mocking into a `beforeEach`?

File: src/notes/notes.service.ts [hunk 5]
```diff
@@ -74,21 +91,56 @@ export class NotesService {
...
   }
 
-  async getCurrentContent(note: Note): Promise<string> {
+  /**
+   * @async
+   * Get the current content of the note.
+   * @param {Note} note - the note to use
+   * @return {string} the content of the note
+   */
+  async getNoteContentByNote(note: Note): Promise<string> {
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/notes/notes.service.ts:114]: There is already a function `getNoteContentByIdOrAlias`

File: src/notes/notes.service.spec.ts [hunk 6]
```diff
@@ -67,9 +79,803 @@ describe('NotesService', () => {
...
+    groupPermissionUpate.canEdit = false;
+    const user = User.create(userPermissionUpdate.username, 'Testy') as User;
+    const group = Group.create(
+      groupPermissionUpate.groupname,
+      groupPermissionUpate.groupname,
+    );
+    const note = Note.create(user);
+    describe('works', () => {
+      it('with empty GroupPermissions and with empty UserPermissions', async () => {
+        jest.spyOn(noteRepo, 'findOne').mockResolvedValueOnce(note);
```
Comment (davidmehren) [inline src/notes/notes.service.spec.ts]: Maybe move the mocking into a `beforeEach`?

File: src/users/users.service.ts [hunk 7]
```diff
@@ -63,7 +63,7 @@ export class UsersService {
       userName: user.userName,
       displayName: user.displayName,
       photo: this.getPhotoUrl(user),
-      email: user.email,
+      email: user.email ?? '',
```
Comment (davidmehren) [inline src/users/users.service.ts:66]: Should  we make `email` optional in `UserInfoDto` instead?

`null` is probably clearer than an empty string, when the user has no mail address.

**Review by davidmehren — CHANGES_REQUESTED** [review]

**Review by InnayTool — APPROVED** [review]

File: src/notes/notes.service.ts [hunk 8]
```diff
@@ -82,7 +82,6 @@ export class NotesService {
   ): Promise<Note> {
     const newNote = Note.create();
     newNote.revisions = Promise.resolve([
-      //TODO: Calculate patch
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/notes/notes.service.ts:67]: fixed

File: src/notes/notes.service.ts [hunk 9]
```diff
@@ -215,44 +216,114 @@ export class NotesService {
...
+    const groups = newPermissions.sharedToGroups.map(
+      (groupPermission) => groupPermission.groupname,
+    );
+
+    if (
+      checkArrayForDuplicates(users) ||
+      checkArrayForDuplicates(groups)
+    ) {
+      this.logger.debug(
+        `The PermissionUpdate '${newPermissions}' you requested specifies the same user or group multiple times.`,
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/notes/notes.service.ts:230]: fixed

File: src/notes/notes.service.ts [hunk 10]
```diff
@@ -215,44 +216,90 @@ export class NotesService {
...
+  /**
+   * @async
+   * Calculate the updateUser (for the NoteDto) for a Note.
+   * @param {Note} note - the note to use
+   * @return {User} user to be used as updateUser in the NoteDto
+   */
+  async calculateUpdateUser(note: Note): Promise<User> {
+    const lastRevision = await this.getLatestRevision(note);
+    if (lastRevision && lastRevision.authorships) {
+      // Sort the last Revisions AuthorShips by their updatedAt Date to get the latest one
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/notes/notes.service.ts:288]: fixed

File: src/notes/notes.service.spec.ts (hunk 3 above)
Comment (DerMolly) [reply src/notes/notes.service.spec.ts:100]: fixed

File: src/notes/notes.service.spec.ts (hunk 4 above)
Comment (DerMolly) [reply src/notes/notes.service.spec.ts:121]: fixed

File: src/notes/notes.service.spec.ts (hunk 6 above)
Comment (DerMolly) [reply src/notes/notes.service.spec.ts]: This is not possible as the resolved value is not the same in all tests…

File: src/users/users.service.ts (hunk 7 above)
Comment (DerMolly) [reply src/users/users.service.ts:66]: This was (as I already stated in the commit message) in an effort to be consistent with photoUrl…

**Review by InnayTool — APPROVED** [review]

**Review by davidmehren — APPROVED** [review]

**Pull request merged**